### PR TITLE
[Swift3] Make extracting info from errors easier.

### PR DESCRIPTION
**Pull request by GriffinSchneider**
This pull request makes 2 changes to how errors are handled by Swift3 clients.

- Error handlers used to define their error argument as `Error?`, when it was actually always a `ErrorResponse?`. Now errors are typed as `ErrorResponse?` so there's no need to cast them all the time.
- `ErrorResponse` used to be an enum with only 1 value. Now it's a struct so there's no need to do a `switch` with only 1 `case` every time you want to get a property of the error.

For example, let's say you want to make some request to your API and then print out the status code if there's an error. 

**Before** this pull request, you'd have to do something like this:
```
DefaultAPI.someRequest() { res, error in
    if let error = error {
        let err = error as! ErrorResponse 
        switch(err) {
        case let .Error(statusCode, data, error):
            print(statusCode)
        }
        return
    }
    // There was no error
}
```

**After** this pullreq, it's just this:

```
DefaultAPI.someRequest() { res, error in
    if let error = error {
        print(error.statusCode)
        return
    }
    // There was no error
}
```

### CODE REVIEW

**Comment by jgavris** [comment]
I was actually working on something similar myself. This will be a breaking change so I believe this should be pointed at the `2.3.0` branch.

File: modules/swagger-codegen/src/main/resources/swift3/Models.mustache [hunk 1]
```diff
@@ -10,8 +10,10 @@ protocol JSONEncodable {
     func encodeToJSON() -> Any
 }
 
-public enum ErrorResponse : Error {
-    case Error(Int, Data?, Error)
+public struct ErrorResponse : Error {
+    public let statusCode: Int
+    public let data: Data?
+    public let error: Error
 }
```
Comment (jgavris) [inline modules/swagger-codegen/src/main/resources/swift3/Models.mustache]: Would also be lovely to implement `CustomStringConvertible` and print out the response data as a string.

**Comment by jaz-ah** [comment]
@wing328 for breaking changes do they go to master right now or do we push to 2.3.0?

**Comment by jgavris** [comment]
@jaz-ah It's in the PR template notes 

>- [ ] Filed the PR against the correct branch: master for non-breaking changes and `2.3.0` branch for breaking (non-backward compatible) changes.

That being said, [semantic versioning](http://semver.org) would suggest that breaking changes should involve a major version bump.

**Comment by wing328** [comment]
For breaking change, please file against 2.3.0 branch instead.

**Comment by GriffinSchneider** [comment]
Ok, I changed this pull request to point at `2.3.0` .

The CI fails now, but the Travis CI failure seems like it isn't swift-related:
```[INFO] TS Node npm Petstore Client ....................... FAILURE [10.548s]```
so I'm not sure what to do about it.

File: modules/swagger-codegen/src/main/resources/swift3/Models.mustache (hunk 1 above)
Comment (GriffinSchneider) [inline modules/swagger-codegen/src/main/resources/swift3/Models.mustache]: Even without making this `CustomStringConvertible`, Swift's default string conversion for structs will still let you see all the data - if you have a suggestion for a better format though, I'd be happy to implement it.

**Comment by wing328** [comment]
Test results look good:
```
** TEST SUCCEEDED **

Test Suite 'All tests' started at 2017-01-10 23:46:58.651
Test Suite 'SwaggerClientTests.xctest' started at 2017-01-10 23:46:58.652
Test Suite 'PetAPITests' started at 2017-01-10 23:46:58.653
Test Case '-[SwaggerClientTests.PetAPITests test1CreatePet]' started.
Test Case '-[SwaggerClientTests.PetAPITests test1CreatePet]' passed (1.158 seconds).
Test Case '-[SwaggerClientTests.PetAPITests test2GetPet]' started.
Test Case '-[SwaggerClientTests.PetAPITests test2GetPet]' passed (0.563 seconds).
Test Case '-[SwaggerClientTests.PetAPITests test3DeletePet]' started.
Test Case '-[SwaggerClientTests.PetAPITests test3DeletePet]' passed (0.505 seconds).
Test Suite 'PetAPITests' passed at 2017-01-10 23:47:00.885.
	 Executed 3 tests, with 0 failures (0 unexpected) in 2.226 (2.232) seconds
Test Suite 'StoreAPITests' started at 2017-01-10 23:47:00.887
Test Case '-[SwaggerClientTests.StoreAPITests test1PlaceOrder]' started.
Test Case '-[SwaggerClientTests.StoreAPITests test1PlaceOrder]' passed (0.520 seconds).
Test Case '-[SwaggerClientTests.StoreAPITests test2GetOrder]' started.
Test Case '-[SwaggerClientTests.StoreAPITests test2GetOrder]' passed (0.509 seconds).
Test Case '-[SwaggerClientTests.StoreAPITests test3DeleteOrder]' started.
Test Case '-[SwaggerClientTests.StoreAPITests test3DeleteOrder]' passed (0.502 seconds).
Test Case '-[SwaggerClientTests.StoreAPITests testDownloadProgress]' started.
Test Case '-[SwaggerClientTests.StoreAPITests testDownloadProgress]' passed (0.477 seconds).
Test Suite 'StoreAPITests' passed at 2017-01-10 23:47:02.904.
	 Executed 4 tests, with 0 failures (0 unexpected) in 2.009 (2.017) seconds
Test Suite 'UserAPITests' started at 2017-01-10 23:47:02.905
Test Case '-[SwaggerClientTests.UserAPITests test1CreateUser]' started.
Test Case '-[SwaggerClientTests.UserAPITests test1CreateUser]' passed (0.494 seconds).
Test Case '-[SwaggerClientTests.UserAPITests test2GetUser]' started.
Test Case '-[SwaggerClientTests.UserAPITests test2GetUser]' passed (0.499 seconds).
Test Case '-[SwaggerClientTests.UserAPITests test3DeleteUser]' started.
Test Case '-[SwaggerClientTests.UserAPITests test3DeleteUser]' passed (0.472 seconds).
Test Case '-[SwaggerClientTests.UserAPITests testLogin]' started.
Test Case '-[SwaggerClientTests.UserAPITests testLogin]' passed (0.532 seconds).
Test Case '-[SwaggerClientTests.UserAPITests testLogout]' started.
Test Case '-[SwaggerClientTests.UserAPITests testLogout]' passed (0.480 seconds).
Test Suite 'UserAPITests' passed at 2017-01-10 23:47:05.385.
	 Executed 5 tests, with 0 failures (0 unexpected) in 2.476 (2.481) seconds
Test Suite 'SwaggerClientTests.xctest' passed at 2017-01-10 23:47:05.386.
	 Executed 12 tests, with 0 failures (0 unexpected) in 6.710 (6.734) seconds
Test Suite 'All tests' passed at 2017-01-10 23:47:05.388.
	 Executed 12 tests, with 0 failures (0 unexpected) in 6.710 (6.737) seconds
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:15 min
[INFO] Finished at: 2017-01-10T23:47:05+08:00
[INFO] Final Memory: 11M/245M
[INFO] ------------------------------------------------------------------------
```

**Pull request merged**